### ISSUE TITLE
Facilitate per-group metrics

### ISSUE THREAD

**Issue by ClaytonJY** [tidymodels/yardstick#20]
I often find myself having a large tibble of predictions from different models. If I want to use `yardstick` to compute metrics per-model, the experience is a bit rough:

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(yardstick)
#> Loading required package: broom
#> For binary classification, the first factor level is assumed to be the event.
#> Set the global option `yardstick.event_first` to `FALSE` to change this.

tbl <- tibble(
  model_id  = rep(1:2, each = 3),
  actual    = rep(1:3, 2),
  predicted = c(1, 1, 3, 3, 2, 1)
) %>%
  mutate_at(vars(-model_id), factor, levels = 1:3)

# no grouping
accuracy(tbl, actual, predicted)
#> [1] 0.5

# grouping ignored
tbl %>%
  group_by(model_id) %>%
  accuracy(actual, predicted)
#> [1] 0.5

# works but ugly
tbl %>%
  group_by(model_id) %>%
  do(log_loss = accuracy(., actual, predicted)) %>%
  ungroup() %>%
  mutate(log_loss = rlang::flatten_dbl(log_loss))
#> # A tibble: 2 x 2
#>   model_id log_loss
#>      <int>    <dbl>
#> 1        1    0.667
#> 2        2    0.333


# example implementation
accuracy.grouped_df <- function(tbl, actual, predicted, out = "accuracy") {
  actual <- enquo(actual)
  predicted <- enquo(predicted)
  
  # wouldn't use MLmetrics in actual implementation
  summarize(tbl, !!out := MLmetrics::Accuracy(predicted, actual))
}

# works now!
tbl %>%
  group_by(model_id) %>%
  accuracy(actual, predicted)
#> # A tibble: 2 x 2
#>   model_id accuracy
#>      <int>    <dbl>
#> 1        1    0.667
#> 2        2    0.333
```

Created on 2018-06-07 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).
<details><summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.4 (2018-03-15)
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2018-06-07
#> Packages -----------------------------------------------------------------
#>  package    * version date       source                               
#>  assertthat   0.2.0   2017-04-11 CRAN (R 3.4.3)                       
#>  backports    1.1.2   2017-12-13 CRAN (R 3.4.3)                       
#>  base       * 3.4.4   2018-03-16 local                                
#>  bindr        0.1.1   2018-03-13 CRAN (R 3.4.4)                       
#>  bindrcpp   * 0.2.2   2018-03-29 cran (@0.2.2)                        
#>  broom      * 0.4.4   2018-03-29 CRAN (R 3.4.4)                       
#>  cli          1.0.0   2017-11-05 CRAN (R 3.4.3)                       
#>  colorspace   1.3-2   2016-12-14 CRAN (R 3.4.3)                       
#>  compiler     3.4.4   2018-03-16 local                                
#>  crayon       1.3.4   2017-09-16 CRAN (R 3.4.3)                       
#>  datasets   * 3.4.4   2018-03-16 local                                
#>  devtools     1.13.5  2018-02-18 CRAN (R 3.4.4)                       
#>  digest       0.6.15  2018-01-28 CRAN (R 3.4.3)                       
#>  dplyr      * 0.7.5   2018-05-19 CRAN (R 3.4.4)                       
#>  evaluate     0.10.1  2017-06-24 CRAN (R 3.4.3)                       
#>  foreign      0.8-70  2018-04-23 CRAN (R 3.4.4)                       
#>  ggplot2      2.2.1   2016-12-30 CRAN (R 3.4.3)                       
#>  glue         1.2.0   2017-10-29 CRAN (R 3.4.3)                       
#>  graphics   * 3.4.4   2018-03-16 local                                
#>  grDevices  * 3.4.4   2018-03-16 local                                
#>  grid         3.4.4   2018-03-16 local                                
#>  gtable       0.2.0   2016-02-26 CRAN (R 3.4.3)                       
#>  htmltools    0.3.6   2017-04-28 CRAN (R 3.4.3)                       
#>  knitr        1.20.3  2018-05-17 Github (yihui/knitr@dc028f4)         
#>  lattice      0.20-35 2017-03-25 CRAN (R 3.3.3)                       
#>  lazyeval     0.2.1   2017-10-29 CRAN (R 3.4.3)                       
#>  magrittr     1.5     2014-11-22 CRAN (R 3.4.3)                       
#>  memoise      1.1.0   2017-04-21 CRAN (R 3.4.3)                       
#>  methods    * 3.4.4   2018-03-16 local                                
#>  MLmetrics    1.1.1   2016-05-13 cran (@1.1.1)                        
#>  mnormt       1.5-5   2016-10-15 CRAN (R 3.4.3)                       
#>  munsell      0.4.3   2016-02-13 CRAN (R 3.4.3)                       
#>  nlme         3.1-137 2018-04-07 CRAN (R 3.4.4)                       
#>  parallel     3.4.4   2018-03-16 local                                
#>  pillar       1.2.3   2018-05-25 CRAN (R 3.4.4)                       
#>  pkgconfig    2.0.1   2017-03-21 CRAN (R 3.4.3)                       
#>  plyr         1.8.4   2016-06-08 CRAN (R 3.4.3)                       
#>  pROC         1.12.1  2018-05-06 CRAN (R 3.4.4)                       
#>  psych        1.8.4   2018-05-06 CRAN (R 3.4.4)                       
#>  purrr        0.2.5   2018-05-29 CRAN (R 3.4.4)                       
#>  R6           2.2.2   2017-06-17 CRAN (R 3.4.3)                       
#>  Rcpp         0.12.17 2018-05-18 CRAN (R 3.4.4)                       
#>  reshape2     1.4.3   2017-12-11 CRAN (R 3.4.3)                       
#>  rlang        0.2.1   2018-05-30 CRAN (R 3.4.4)                       
#>  rmarkdown    1.9     2018-03-01 CRAN (R 3.4.4)                       
#>  rprojroot    1.3-2   2018-01-03 CRAN (R 3.4.3)                       
#>  scales       0.5.0   2017-08-24 CRAN (R 3.4.3)                       
#>  stats      * 3.4.4   2018-03-16 local                                
#>  stringi      1.2.2   2018-05-02 CRAN (R 3.4.4)                       
#>  stringr      1.3.1   2018-05-10 CRAN (R 3.4.4)                       
#>  tibble       1.4.2   2018-01-22 CRAN (R 3.4.3)                       
#>  tidyr        0.8.1   2018-05-18 CRAN (R 3.4.4)                       
#>  tidyselect   0.2.4   2018-02-26 CRAN (R 3.4.4)                       
#>  tools        3.4.4   2018-03-16 local                                
#>  utf8         1.1.4   2018-05-24 CRAN (R 3.4.4)                       
#>  utils      * 3.4.4   2018-03-16 local                                
#>  withr        2.1.2   2018-03-15 CRAN (R 3.4.4)                       
#>  yaml         2.1.19  2018-05-01 CRAN (R 3.4.4)                       
#>  yardstick  * 0.0.1   2018-06-07 Github (tidymodels/yardstick@3432d97)
```

</details>

One potential way forward would be like I did above with a method for "grouped_df", maybe even using the `do` approach instead, though that feels a bit messy.

Something that might work nice is to have the default methods expect vector inputs (like `MLmetrics` functions), but give them `tbl_df`/`grouped_df`/`matrix` methods which call the default and add the syntactical sugar on top. Comparing with the current implementation, this would essentially have default/vector methods replace `table` methods as the lowest-common-denominator the other methods call underneath.

The syntactic sugar required for e.g. `data.frame` methods might also be consolidated further with an additional layer of abstraction to handle e.g. two-class/class-probability/regression/etc families.

If adding this functionality makes sense, I can help!

**Comment by DavisVaughan** [comment]
I'm working on this over in the branch `feature/yardstick-rewrite-for-groups-vecs-and-extensibility` if you want to check it out. I've got it working for everything in `regression.R` (docs aren't updated fully yet).

We are changing `rmse()` to return a 1 row tibble, so that the output of `rmse()` on a grouped df returns a tibble with as many rows as there are groups. `rmse_vec()` takes vectors and returns 1 number like the old version.

Also check out `metric_set()` for a way to create bind multiple metrics together.

```r
devtools::install_github("tidymodels/yardstick", ref = "feature/yardstick-rewrite-for-groups-vecs-and-extensibility")
```

``` r
library(dplyr)
library(yardstick)

tbl <- tibble(
  model_id  = rep(1:2, each = 3),
  actual    = rnorm(6, mean = 1, sd = .01),
  predicted = rnorm(6, mean = 1, sd = .01)
)

# non grouped now returns a tibble
tbl %>%
  rmse(actual, predicted)
#> # A tibble: 1 x 2
#>   name    value
#>   <chr>   <dbl>
#> 1 rmse  0.00726

# groups no longer ignored!
# (non grouped version returns a tibble to be consistent with this)
tbl %>%
  group_by(model_id) %>%
  rmse(actual, predicted)
#> # A tibble: 2 x 3
#>   model_id name    value
#>      <int> <chr>   <dbl>
#> 1        1 rmse  0.00715
#> 2        2 rmse  0.00737

# vector versions for old behavior
rmse_vec(tbl$actual, tbl$predicted)
#> [1] 0.007257038

tbl %>%
  group_by(model_id) %>%
  mutate(val = rmse_vec(actual, predicted))
#> # A tibble: 6 x 4
#> # Groups:   model_id [2]
#>   model_id actual predicted     val
#>      <int>  <dbl>     <dbl>   <dbl>
#> 1        1  1.01      1.00  0.00715
#> 2        1  0.993     0.992 0.00715
#> 3        1  1.01      1.000 0.00715
#> 4        2  0.992     0.987 0.00737
#> 5        2  1.00      1.01  0.00737
#> 6        2  0.992     0.982 0.00737

# Metric sets for user defined groups of metrics
# (this is why we now return a "tidy" 1 row data frame for each metric. stackability!)
multi_metric <- metric_set(rmse, rsq, rsq_trad)

tbl %>%
  group_by(model_id) %>%
  multi_metric(actual, predicted)
#> # A tibble: 6 x 3
#>   model_id name        value
#>      <int> <chr>       <dbl>
#> 1        1 rmse      0.00715
#> 2        2 rmse      0.00737
#> 3        1 rsq       0.936  
#> 4        2 rsq       0.982  
#> 5        1 rsq_trad  0.219  
#> 6        2 rsq_trad -0.937
```

Created on 2018-09-11 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by ClaytonJY** [comment]
@DavisVaughan very cool, thanks for sharing. Curious to see how that plays out in the weirder world of classification & probability where the args vary more. Can I make a metric set of accuracy and log-loss?

I'm a little torn on the `metric_set` output; that long format is what I want for plotting, but for every other use (display, saving to disk) I'd rather have a column for each metric (one row per group). Is there an expectation of that comes next in the common/recommended workflow, does that inform the choice of long vs. wide?

On the `*_vec`, would it be possible to have a default method instead, so the functions just "do the right thing" based on if the first arg is a vector or data.frame?

**Comment by DavisVaughan** [comment]
Averaging metrics across folds is a case that comes to mind where there is still processing to do after calculating the metrics on a per-fold basis. 

See the group_by(epoch) and summarise() call in the article below. The tidy long format lends itself to easily doing this for many metrics at once

https://tidymodels.github.io/rsample/articles/Applications/Keras.html

**Comment by ClaytonJY** [comment]
Er, bringing up `metric_set` there confuses the point, because that isn't how you'd likely handle one model. I think a better way towards my point is, if you also wanted to measure MAPE in addition to RMSE for each model from `mlp_rmse`, would you want it to be another _column_ or another _row_?

**Comment by DavisVaughan** [comment]
[rsample-keras-rewrite.zip](https://github.com/tidymodels/yardstick/files/2379342/rsample-keras-rewrite.zip)

I've reworked the keras example to work with the new version of `yardstick`. This is what i imagine the flow would look like. It looks pretty smooth to me tbh. I do think that when I want to add another measure, I want it in a new row because downstream operations are easier with more groups rather than more columns.

**Comment by DavisVaughan** [comment]
> On the *_vec, would it be possible to have a default method instead, so the functions just "do the right thing" based on if the first arg is a vector or data.frame?

On this front, I have two reasons for wanting to keep `*_vec()` functions.

1) signature differences. Below, it makes sense to have `data` as the first arg for `data.frame`s but what about for vectors? I feel like it makes sense to promote `truth` and `estimate` to the only args there because they match 1:1 with what is passed into the data frame version. An alternative is `rmse(x, truth, estimate)` for data frames and `rmse(x, estimate)` for vectors, but now you have the truth variable being `truth` in the data frame version and `x` in the vector version, and I don't like that.

```r
> rmse.data.frame
function(data, truth, estimate, na.rm = TRUE, ...)
```

```r
> rmse_vec
function(truth, estimate, na.rm = TRUE, ...)
```

2) return type. `rmse()` is always going to return a tibble, whether the input is `data.frame` `matrix` or a `table`. `rmse_vec()` is always going to return a vector. I want to retain that "I expect to always get xyz from this function" mindset so I don't want to have `rmse()` return tibbles sometimes and vectors other times depending on the input.
